app/lib/katello/resources/discovery/container_discovery.rb

```diff
@@ -0,0 +1,123 @@
+      @to_follow = to_follow
+    end
+
+    def run(_resume_point)
+      docker_search
+    end
+
+    private
+
+    def parse_parameter(field_value)
+      field_value.lstrip!
+      i = field_value.index(/[ \t=;,]/) || field_value.length
+      name = field_value.slice!(0, i).downcase(:ascii)
+      field_value.lstrip!
+      if field_value.delete_prefix!('=')
+        field_value.lstrip!
+        if field_value.delete_prefix!('"')
+          value = ''
+          until field_value.empty?
+            break if field_value.delete_prefix!('"')
+            field_value.delete_prefix!("\\")
+            value += field_value.slice!(0, 1) || ''
+          end
+        else
+          i = field_value.index(/[;,]/) || field_value.length
+          value = field_value.slice!(0, i)
+        end
+      end
+      {name: name, value: value || ''}
+    end
+
+    def parse_parameters(field_value)
+      seen_rel = false
+      has_next_rel = false
+      has_anchor = false
+      until field_value.empty?
+        field_value.lstrip!
+        break if field_value.delete_prefix!(';').nil?
+        param = parse_parameter(field_value)
+        case
+        when param[:name] == 'rel' && !seen_rel
+          seen_rel = true
+          has_next_rel = param[:value].downcase(:ascii).split(/[ \t]/).include?('next')
+        when param[:name] == 'anchor'
+          has_anchor = true
+        end
+      end
+      {has_next_rel: has_next_rel, has_anchor: has_anchor}
+    end
+
+    def get_next_link(link_header)
+      # This code mostly follows Appendix B "Algorithms of Parsing Link Header Fields" of RFC 8288
+      # "Web Linking", <https://www.rfc-editor.org/rfc/rfc8288#appendix-B> (that RFC appears to be
+      # silent about multiple "next" links, so just use the first one and ignore any additional
+      # ones, in the general spirit of being lenient):
+      return nil if link_header.nil?
+      field_value = link_header.clone
+      until field_value.empty?
+        # The following ignores any junk preceding the next <...> link URL:
+        m = field_value.match(/<(.*)>/)
+        break unless m
+        target_string = m[1]
+        field_value = m.post_match
+        params = parse_parameters(field_value)
+        if params[:has_next_rel]
+          # To keep it simple, ignore a link with an (unlikely) anchor parameter; but the RFC
+          # mandates that we "MUST NOT process the link without applying the anchor", so just raise
+          # an exception in that (unlikely) case:
+          fail "anchor not supported" if params[:has_anchor]
+          return target_string
+        end
+      end
+      nil
+    end
+
+    def docker_search
+      request_params = {
+        method: :get,
+        headers: { accept: :json },
+        url: "#{@uri}v1/search?q=#{@search}"
+      }
+
+      request_params[:user] = @upstream_username unless @upstream_username.empty?
+      request_params[:password] = @upstream_password unless @upstream_password.empty?
```

Comment:
It's already `nil` or a non-empty string (constructor guarantees that) so why not just:
```suggestion
      request_params[:user] = @upstream_username if @upstream_username
      request_params[:password] = @upstream_password if @upstream_password
```